File: RmlUi.Net/Element.cs
using System.Runtime.InteropServices;

namespace ChickenWithLips.RmlUi;

public interface Element
{
    public IntPtr NativePtr { get; }

    /// <summary>
    /// Gets the name of the element.
    /// </summary>
    public string TagName { get; }

    /// <summary>
    /// Gets the document this element belongs to.
    /// </summary>
    public ElementDocument OwnerDocument { get; }

    /// <summary>
    /// Adds an event listener to this element.
    /// </summary>
    /// <param name="name">The name of the event to attach to.</param>
    /// <param name="eventListener">Listener object to be attached.</param>
    /// <param name="inCapturePhase">True if the listener is to be attached to the capture phase, false for the bubble phase.</param>
    public void AddEventListener(string name, EventListener eventListener, bool inCapturePhase = false);

    /// <summary>
    /// Adds an event listener to this element.
    /// </summary>
    /// <param name="name">The name of the event to attach to.</param>
    /// <param name="eventListener">Listener object to be attached.</param>
    /// <param name="inCapturePhase">True if the listener is to be attached to the capture phase, false for the bubble phase.</param>
    public void RemoveEventListener(string name, EventListener eventListener, bool inCapturePhase = false);

    /// <summary>
    /// Get a child element by its ID.
    /// </summary>
    /// <param name="id">ID of the the child element</param>
    /// <returns>The child of this element with the given ID, or null if no such child exists.</returns>
    public Element? GetElementById(string id);

    /// <summary>
    /// Sets the markup and content of the element. All existing children will be replaced.
    /// </summary>
    /// <param name="rml">The new content of the element.</param>
    public void SetInnerRml(string rml);

    /// <summary>
    /// Gives focus to the current element.
    /// </summary>
    /// <returns>True if the change focus request was successful.</returns>
    public bool Focus();

    /// <summary>
    /// Removes focus from from this element.
    /// </summary>
    public void Blur();

    /// <summary>
    /// Fakes a mouse click on this element.
    /// </summary>
    public void Click();
}

public abstract class Element<T> : RmlBase<T>, Element
    where T : class
{
    #region Properties

    public string TagName => Marshal.PtrToStringAuto(Native.Element.GetTagName(NativePtr));

    /// <summary>
    /// Gets the document this element belongs to.
    /// </summary>
    public ElementDocument OwnerDocument {
        get {
            var elementType = Marshal.PtrToStringAuto(
                Native.Element.GetOwnerDocument(NativePtr, out var elementPtr)
            );

            return Util.GetOrThrowElementByTypeName(elementPtr, elementType) as ElementDocument;
        }
    }

    #endregion

    #region Methods

    protected Element(IntPtr ptr, bool automaticallyRegisterInCache)
        : base(ptr, automaticallyRegisterInCache)
    {
    }

    /// <summary>
    /// Adds an event listener to this element.
    /// </summary>
    /// <param name="name">The name of the event to attach to.</param>
    /// <param name="eventListener">Listener object to be attached.</param>
    /// <param name="inCapturePhase">True if the listener is to be attached to the capture phase, false for the bubble phase.</param>
    public void AddEventListener(string name, EventListener eventListener, bool inCapturePhase = false)
    {
        Native.Element.AddEventListener(NativePtr, name, eventListener.NativePtr, inCapturePhase);
    }

    /// <summary>
    /// Adds an event listener to this element.
    /// </summary>
    /// <param name="name">The name of the event to attach to.</param>
    /// <param name="eventListener">Listener object to be attached.</param>
    /// <param name="inCapturePhase">True if the listener is to be attached to the capture phase, false for the bubble phase.</param>
    public void RemoveEventListener(string name, EventListener eventListener, bool inCapturePhase = false)
    {
        Native.Element.RemoveEventListener(NativePtr, name, eventListener.NativePtr, inCapturePhase);
    }

    /// <summary>
    /// Get a child element by its ID.
    /// </summary>
    /// <param name="id">ID of the the child element</param>
    /// <returns>The child of this element with the given ID, or null if no such child exists.</returns>
    public Element? GetElementById(string id)
    {
        var nativeElementType = Native.Element.GetElementById(NativePtr, id, out var elementPtr);

        if (null == nativeElementType) {
            return null;
        }

        var elementType = Marshal.PtrToStringAuto(nativeElementType);

        if (null == elementType) {
            return null;
        }

        return Util.GetOrThrowElementByTypeName(elementPtr, elementType);
    }

    /// <summary>
    /// Sets the markup and content of the element. All existing children will be replaced.
    /// </summary>
    /// <param name="rml">The new content of the element.</param>
    public void SetInnerRml(string rml)
    {
        Native.Element.SetInnerRml(NativePtr, rml);
    }

    /// <summary>
    /// Gives focus to the current element.
    /// </summary>
    /// <returns>True if the change focus request was successful.</returns>
    public bool Focus()
    {
        return Native.Element.Focus(NativePtr);
    }

    /// <summary>
    /// Removes focus from from this element.
    /// </summary>
    public void Blur()
    {
        Native.Element.Blur(NativePtr);
    }

    /// <summary>
    /// Fakes a mouse click on this element.
    /// </summary>
    public void Click()
    {
        Native.Element.Click(NativePtr);
    }

    #endregion
}

File: RmlUi.Native/src/Element.cpp
#include "RmlNative.h"
#include "RmlUi/Core/Element.h"
#include "Util.h"

RMLUI_CAPI const char *rml_Element_GetTagName(Rml::Element *element) {
    return element->GetTagName().c_str();
}

RMLUI_CAPI void
rml_Element_AddEventListener(Rml::Element *element, const char *event, Rml::EventListener *eventListener,
                             bool inCapturePhase) {
    element->AddEventListener(event, eventListener, inCapturePhase);
}

RMLUI_CAPI void
rml_Element_RemoveEventListener(Rml::Element *element, const char *event, Rml::EventListener *eventListener,
                                bool inCapturePhase) {
    element->RemoveEventListener(event, eventListener, inCapturePhase);
}

RMLUI_CAPI const char *rml_Element_GetElementByID(Rml::Element *element, const char *id, Rml::Element **foundElement) {
    *foundElement = element->GetElementById(id);

    if (*foundElement) {
        return rmlui_type_name(**foundElement);
    }

    return nullptr;
}

RMLUI_CAPI const char *rml_Element_GetOwnerDocument(Rml::Element *element, Rml::Element **foundElement) {
    *foundElement = element->GetOwnerDocument();

    return rmlui_type_name(**foundElement);
}

RMLUI_CAPI void rml_Element_SetInnerRml(Rml::Element *element, const char* rml) {
    element->SetInnerRML(rml);
}

RMLUI_CAPI void rml_Element_Focus(Rml::Element *element) {
    element->Focus();
}

RMLUI_CAPI void rml_Element_Blur(Rml::Element *element) {
    element->Blur();
}

RMLUI_CAPI void rml_Element_Click(Rml::Element *element) {
    element->Click();
}
